src/lang/ir.cc

```diff
@@ -132,6 +298,802 @@ Expr Any::make() {
+.set_dispatch<AssertStmt>([](const AssertStmt* op, IRPrinter* p) {
+    p->PrintIndent();
+    p->stream << "assert(";
+    p->Print(op->condition);
+    p->stream << ", ";
+    p->Print(op->message);
+    p->stream << ")\n";
+    p->Print(op->body);
+  });
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<ProducerConsumer>([](const ProducerConsumer* op, IRPrinter* p) {
+    if (op->is_producer) {
+      p->PrintIndent();
+      p->stream << "produce " << op->func->func_name() << " {\n";
+      p->indent += 2;
+      p->Print(op->body);
+      p->indent -= 2;
+      p->PrintIndent();
+      p->stream << "}\n";
+    } else {
+      p->Print(op->body);
+    }
+  });
+
+std::ostream &operator<<(std::ostream& out, ForType type) { // NOLINT(*)
+  switch (type) {
+    case ForType::Serial:
+      out << "for";
+      break;
+    case ForType::Parallel:
+      out << "parallel";
+      break;
+    case ForType::Unrolled:
+      out << "unrolled";
+      break;
+    case ForType::Vectorized:
+      out << "vectorized";
+      break;
+  }
+  return out;
+}
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<For>([](const For* op, IRPrinter* p) {
+    p->PrintIndent();
+    p->stream << op->for_type << " (" << op->loop_var << ", ";
+    p->Print(op->min);
+    p->stream << ", ";
+    p->Print(op->extent);
+    p->stream << ") {\n";
+
+    p->indent += 2;
+    p->Print(op->body);
+    p->indent -= 2;
+
+    p->PrintIndent();
+    p->stream << "}\n";
+});
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<Store>([](const Store* op, IRPrinter* p) {
+    p->PrintIndent();
+    p->stream << op->buffer_var << "[";
+    p->Print(op->index);
+    p->stream << "] = ";
+    p->Print(op->value);
+    if (!is_one(op->predicate)) {
+      p->stream << " if ";
+      p->Print(op->predicate);
+    }
+    p->stream << '\n';
+  });
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<Provide>([](const Provide* op, IRPrinter* p) {
+    p->PrintIndent();
+    p->stream << op->func->func_name() << "(";
+    for (size_t i = 0; i < op->args.size(); ++i) {
+      p->Print(op->args[i]);
+      if (i < op->args.size() - 1) p->stream << ", ";
+    }
+    p->stream << ")";
+    if (op->func->num_outputs() != 1) {
+      p->stream << ".value[" << op->value_index << "]";
+    }
+    p->stream << " =";
+    p->Print(op->value);
+    p->stream << '\n';
+  });
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<Allocate>([](const Allocate* op, IRPrinter* p) {
+    p->PrintIndent();
+    p->stream << "allocate " << op->buffer_var << "[" << op->type;
+    for (size_t i = 0; i < op->extents.size(); ++i) {
+      p->stream << " * ";
+      p->Print(op->extents[i]);
+    }
+    p->stream << "]";
+    if (!is_one(op->condition)) {
+      p->stream << " if ";
+      p->Print(op->condition);
+    }
+    p->stream << "\n";
+    p->Print(op->body);
+  });
+
+TVM_STATIC_IR_FUNCTOR(IRPrinter, vtable)
+.set_dispatch<Free>([](const Free* op, IRPrinter* p) {
```

Comment:
I think we should unify both printers and start using a layout algorithm, we already have some lines which are incredibly ugly in current printing strategy, probably out of scope of this PR though.

cc @MarisaKirisame

Reply:
There is a lot of work in the printer. I am still trying to wrap up roundtripping.